jetty-server/src/main/java/org/eclipse/jetty/server/BlockingContentProducer.java

```diff
@@ -101,14 +110,33 @@ public boolean consumeAll(Throwable x)
             if (LOG.isDebugEnabled())
                 LOG.debug("nextContent async producer is not ready, waiting on semaphore {}", _semaphore);
 
+            AutoLock lock = _asyncContentProducer.lock();
+            // Start the counter at -1 b/c acquiring the lock increases the hold count by 1.
+            int lockHoldCount = -1;
             try
             {
+                // Do not hold the lock during the wait on the semaphore;
+                // the lock must be unlocked more than once because it is
+                // reentrant so it can be acquired multiple times by the same
+                // thread, hence it can still be held after a single unlock.
+                while (lock.isHeldByCurrentThread())
+                {
+                    lock.close();
+                    lockHoldCount++;
+                }
```

Comment:
Can this now be done with a Condition rather than a Semaphore?

Reply:
Unfortunately not because the signaling can happen before the reader thread reaches the awaiting on the lock, hence losing the signal to wake up. This is the reason why a `Semaphore` is used: to remember that the signaling happened and that the next await does not need to block. This is the key concept that helped simplify the input state machine.

Reply:
That's true when we were not grabbing a lock, but now that we have a lock we can have a protected int or boolean that we handle like a semaphore:
 + if we are consuming and the int is > 1, then we decrement and continue else we wait on the condition.
 + if we are producing we increment the int, if it was 0 then we signal the condition.

We don't have to release locks then.

Reply:
Good point, I'm going to try this out.

Reply:
Done.